I couldn't compile kyfd

Hello, Graham Neubig.

Thank you for your sharing weighted finite state transducer decoder, I am a researcher who is  interested in finite state transducer technology. 

By Googling, I found your decoder kyfd, 
When I compiled kyfd v 0.0.5 from your site http://www.phontron.com/kyfd/, with recent openfst-1.5.2,
the compiler showed the followings,

I added the option " -std=c++11" to every Makefiles, but I couldn't compile properly,
Please let me know how I can compile your code properly,
Thank you again for your knowledge sharing! 

/bin/bash ../../libtool --tag=CXX   --mode=compile g++ -DHAVE_CONFIG_H -I. -I../..  -I./../include -I/src/bin -std=c++11  -g -O2 -std=c++11 -MT decoder.lo -MD -MP -MF .deps/decoder.Tpo -c -o decoder.lo decoder.cc
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I../.. -I./../include -I/src/bin -std=c++11 -g -O2 -std=c++11 -MT decoder.lo -MD -MP -MF .deps/decoder.Tpo -c decoder.cc -o decoder.o
In file included from decoder.cc:24:0:
/usr/local/include/fst/compose.h: In instantiation of 'fst::ComposeFstImpl<C, F, T>::Weight fst::ComposeFstImpl<C, F, T>::ComputeFinal(fst::ComposeFstImpl<C, F, T>::StateId) [with C = fst::DefaultCacheStorefst::ArcTpl<fst::TropicalWeightTpl<float > >; F = fst::SequenceComposeFilterfst::FallbackMatcher<fst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float > > > >, fst::FallbackMatcherfst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float > > > > >; T = fst::GenericComposeStateTablefst::ArcTpl<fst::TropicalWeightTpl<float >, fst::IntegerFilterState<signed char>, fst::DefaultComposeStateTuple<int, fst::IntegerFilterState<signed char> >, fst::CompactHashStateTable<fst::DefaultComposeStateTuple<int, fst::IntegerFilterState<signed char> >, fst::ComposeHash<fst::DefaultComposeStateTuple<int, fst::IntegerFilterState<signed char> > > > >; fst::ComposeFstImpl<C, F, T>::Weight = fst::TropicalWeightTpl<float>; fst::ComposeFstImpl<C, F, T>::StateId = int]':
decoder.cc:400:1:   required from here
/usr/local/include/fst/compose.h:391:40: error: 'fst::ComposeFstImplfst::DefaultCacheStore<fst::ArcTpl<fst::TropicalWeightTpl<float > >, fst::SequenceComposeFilterfst::FallbackMatcher<fst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float > > > >, fst::FallbackMatcherfst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float > > > > >, fst::GenericComposeStateTablefst::ArcTpl<fst::TropicalWeightTpl<float >, fst::IntegerFilterState<signed char>, fst::DefaultComposeStateTuple<int, fst::IntegerFilterState<signed char> >, fst::CompactHashStateTable<fst::DefaultComposeStateTuple<int, fst::IntegerFilterState<signed char> >, fst::ComposeHash<fst::DefaultComposeStateTuple<int, fst::IntegerFilterState<signed char> > > > > >::M1' has no member named 'Final'
     Weight final1 = matcher1_->Final(s1);
...
/usr/local/include/fst/compose.h:412:51: error: 'fst::ComposeFstImplfst::DefaultCacheStore<fst::ArcTpl<fst::TropicalWeightTpl<float > >, fst::SequenceComposeFilterfst::FallbackMatcher<fst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float > > > >, fst::FallbackMatcherfst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float > > > > >, fst::GenericComposeStateTablefst::ArcTpl<fst::TropicalWeightTpl<float >, fst::IntegerFilterState<signed char>, fst::DefaultComposeStateTuple<int, fst::IntegerFilterState<signed char> >, fst::CompactHashStateTable<fst::DefaultComposeStateTuple<int, fst::IntegerFilterState<signed char> >, fst::ComposeHash<fst::DefaultComposeStateTuple<int, fst::IntegerFilterState<signed char> > > > > >::M2' has no member named 'Priority'
         ssize_t priority2 = matcher2_->Priority(s2);
                                                   ^

Thanks for the report! Unfortunately I don't have time to look at this now,
but if you figure out the reason I'll be happy to have a pull request.

On Fri, May 13, 2016 at 11:35 AM, bigreddeer notifications@github.com
wrote:

> Hello, Graham Neubig.
> 
> Thank you for your sharing weighted finite state transducer decoder, I am
> a researcher who is interested in finite state transducer technology.
> 
> By Googling, I found your decoder kyfd,
> When I compiled kyfd v 0.0.5 from your site http://www.phontron.com/kyfd/,
> with recent openfst-1.5.2,
> the compiler showed the followings,
> 
> I added the option " -std=c++11" to every Makefiles, but I couldn't
> compile properly,
> Please let me know how I can compile your code properly,
> Thank you again for your knowledge sharing!
> 
> /bin/bash ../../libtool --tag=CXX --mode=compile g++ -DHAVE_CONFIG_H -I.
> -I../.. -I./../include -I/src/bin -std=c++11 -g -O2 -std=c++11 -MT
> decoder.lo -MD -MP -MF .deps/decoder.Tpo -c -o decoder.lo decoder.cc
> libtool: compile: g++ -DHAVE_CONFIG_H -I. -I../.. -I./../include
> -I/src/bin -std=c++11 -g -O2 -std=c++11 -MT decoder.lo -MD -MP -MF
> .deps/decoder.Tpo -c decoder.cc -o decoder.o
> In file included from decoder.cc:24:0:
> /usr/local/include/fst/compose.h: In instantiation of
> 'fst::ComposeFstImpl::Weight
> fst::ComposeFstImpl::ComputeFinal(fst::ComposeFstImpl::StateId) [with C =
> fst::DefaultCacheStorefst::ArcTpl<fst::TropicalWeightTpl<float > >; F =
> fst::SequenceComposeFilterfst::FallbackMatcher<fst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float
> 
> > > > > ,
> > > > > fst::FallbackMatcherfst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float
> > > > > 
> > > > > > ; T =
> > > > > > fst::GenericComposeStateTablefst::ArcTpl<fst::TropicalWeightTpl<float >,
> > > > > > fst::IntegerFilterState, fst::DefaultComposeStateTuple >,
> > > > > > fst::CompactHashStateTable >, fst::ComposeHash > > > >;
> > > > > > fst::ComposeFstImpl::Weight = fst::TropicalWeightTpl;
> > > > > > fst::ComposeFstImpl::StateId = int]':
> > > > > > decoder.cc:400:1: required from here
> > > > > > /usr/local/include/fst/compose.h:391:40: error:
> > > > > > 'fst::ComposeFstImplfst::DefaultCacheStore<fst::ArcTpl<fst::TropicalWeightTpl<float
> > > > > > ,
> > > > > > fst::SequenceComposeFilterfst::FallbackMatcher<fst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float
> > > > > > ,
> > > > > > fst::FallbackMatcherfst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float
> > > > > > ,
> > > > > > fst::GenericComposeStateTablefst::ArcTpl<fst::TropicalWeightTpl<float >,
> > > > > > fst::IntegerFilterState, fst::DefaultComposeStateTuple >,
> > > > > > fst::CompactHashStateTable >, fst::ComposeHash > > > > >::M1' has no member
> > > > > > named 'Final'
> > > > > > Weight final1 = matcher1_->Final(s1);
> > > > > > ...
> > > > > > /usr/local/include/fst/compose.h:412:51: error:
> > > > > > 'fst::ComposeFstImplfst::DefaultCacheStore<fst::ArcTpl<fst::TropicalWeightTpl<float
> > > > > > ,
> > > > > > fst::SequenceComposeFilterfst::FallbackMatcher<fst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float
> > > > > > ,
> > > > > > fst::FallbackMatcherfst::Matcher<fst::Fst<fst::ArcTpl<fst::TropicalWeightTpl<float
> > > > > > ,
> > > > > > fst::GenericComposeStateTablefst::ArcTpl<fst::TropicalWeightTpl<float >,
> > > > > > fst::IntegerFilterState, fst::DefaultComposeStateTuple >,
> > > > > > fst::CompactHashStateTable >, fst::ComposeHash > > > > >::M2' has no member
> > > > > > named 'Priority'
> > > > > > ssize_t priority2 = matcher2_->Priority(s2);
> > > > > > ^
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> https://github.com/neubig/kyfd/issues/1

--- a/src/include/kyfd/fallback-matcher.h
+++ b/src/include/kyfd/fallback-matcher.h
@@ -33,7 +33,7 @@ namespace fst {
 // A heirarchical failure transition model that allows multiple levels of
 //  fallback for phi-transitions
 template <class M>
-class FallbackMatcher {
+class FallbackMatcher: public MatcherBase<class M::Arc> {
  public:
     typedef typename M::FST FST;
     typedef typename M::Arc Arc;

